### PR TITLE
tidy up `Slot`s in the compiler

### LINKED ISSUE #27202: Make Core.Slot a Union type instead of a generic abstract type
For the moment, Core.Slot is defined in src/jltypes.c with 
```
jl_abstractslot_type = jl_new_abstracttype((jl_value_t*)jl_symbol("Slot"), core, jl_any_type,
                                               jl_emptysvec);
```

This type is only ever used in the compiler as a supertype of SlotNumber and TypedSlot, and widely so in queries of the form `isa(x, Slot)`. Since I don't believe it is supposed to be extended with any other subtype, I think it should be defined as `Union{SlotNumber, TypedSlot}` instead of its current definition. 
The main interest of this is that it would allow the compiler to union-split after all the queries `isa(x, Slot)`, which is impossible in the current setting.

As a sidenote, maybe its definition should also be added to the commented part of base/boot.jl, as well as that of SlotNumber and TypedSlot.

### PR DESCRIPTION
**Pull request by aviatesk**
This PR consists of two commits:
- remove `Slot` type from the compiler: since only `SlotNumber` appears in lowering and `TypedSlot` only temporarily appears in a specific phase of inference, there is no need to define abstract type for `SlotNumber` and `TypedSlot`
- move `TypedSlot` definition into `Core.Compiler`: only Julia-level compiler needs to deal with the type

@nanosoldier `runbenchmarks("inference", vs=":master")`

closes #27202

### CODE REVIEW

File: stdlib/Serialization/src/Serialization.jl [hunk 1]
```diff
@@ -39,7 +39,7 @@ const TAGS = Any[
     Float16, Float32, Float64, Char, DataType, Union, UnionAll, Core.TypeName, Tuple,
     Array, Expr, LineNumberNode, :__LabelNode__, GotoNode, QuoteNode, CodeInfo, TypeVar,
     Core.Box, Core.MethodInstance, Module, Task, String, SimpleVector, Method,
-    GlobalRef, SlotNumber, TypedSlot, NewvarNode, SSAValue,
```
Comment (aviatesk) [inline stdlib/Serialization/src/Serialization.jl:42]: Hmm, it looks like we need to replace this `TypedSlot` with something defined in `Core`?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/48693) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/bd5fdc3_vs_d7cbf39/report.md).

File: stdlib/Serialization/src/Serialization.jl (hunk 1 above)
Comment (vtjnash) [reply stdlib/Serialization/src/Serialization.jl:42]: We could just convert to a `SlotNumber` on load from this old tag id (ignore the typ), though I don't think anyone would be using this

**Review by vtjnash — APPROVED** [review]

File: stdlib/Serialization/src/Serialization.jl (hunk 1 above)
Comment (vtjnash) [reply stdlib/Serialization/src/Serialization.jl:42]: I think you need to at least keep a placeholder here though to keep the numbering constant

Comment (aviatesk) [reply stdlib/Serialization/src/Serialization.jl:42]: Can `Const` for example be a placeholder here? Or is something other better?

**Comment by giordano** [comment]
> closes #27202.

Leaving a comment like that does nothing, you probably want to add it to the first message of the PR, so that merging this PR will also automatically close the issue.

**Comment by maleadt** [comment]
This change also broke Traceur and StableHashes. See the latest daily: https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2023-02/17/report.html, lots of packages started failing today. IMO it would have been good to run PkgEval here and try to fix some packages before merging (it will take quite a while now before PkgEval recovers, given that we only test released versions).

**Pull request merged**